### ISSUE TITLE
[Question] How to get consumed power from socket

### ISSUE THREAD

**Issue by arexms** [jasonacox/tinytuya#282]
Hello everyone,
This is my first attempt to get some basic data from device via Tuya API.

I'm able to connect to device and get status data including dps status.
But there is one thing I can't figure out.

Shortly: how to get consumed power from the device?

On the mobile app I can see total (KWh), daily, monthly...  but I didn't find such parameter over API.
Can you help?
My product Type: Socket

![image](https://user-images.githubusercontent.com/18706348/218278694-c1b8da7a-2e4b-44ed-a966-40e28836a5af.png)

**Comment by jasonacox** [comment]
Hi @arexms ,

This question comes up quite a bit. The app uses the Tuya Cloud to compute the energy usage over time (kWh).  The Tuya devices themselves do not seem to compute or store this data, though some do have short term memory to send a "pulse" once they reach some threshold (e.g. 0.1kWh). @make-all does a good job explaining it here: https://github.com/jasonacox/tinytuya/issues/274#issuecomment-1413910082

The DPS values from your device look like they track to this:
https://github.com/jasonacox/tinytuya#version-33---plug-switch-power-strip-type
The one that you are wanting is DPS 17 (Add Electricity) which doesn't show up until it reaches a threshold as mentioned. 

The typical approach is to compute energy (W * t) by sampling the wattage over time and compute the area under the cover (see Riemann sum for approximation using samples).

**Comment by jasonacox** [comment]
What energy monitoring plug did you get @uzlonewolf ? I want to pick one up. The 2 different types I have didn't respond to this, sadly, but then again they only emit when DPS(17) reaches 1.  

FYI - a tiny test loop for fun... if only we had a way of categorizing all our Tuya devices and saying "loop through all devices where type=energy-plugs" 😁 
```python

import tinytuya

plugs = []
plugs.append({'name': 'Type 1','id': 'ebabcdefg1234567890aaa', 'ip': '10.0.1.21', 'version': 3.3})
plugs.append({'name': 'Type 2','id': 'ebabcdefg1234567890bbb', 'ip': '10.0.1.22', 'version': 3.3})
plugs.append({'name': 'Type 3','id': 'ebabcdefg1234567890ccc', 'ip': '10.0.1.23', 'version': 3.4})

for plug in plugs:
    print("\nLooking at %s [%s @ %s : %s: " % (plug['name'], plug['id'],plug['ip'],plug['version']))
    d = tinytuya.OutletDevice(plug['id'],plug['ip'],version=plug['version'])
    d.set_socketPersistent( True )
    print( d.status() )
    d.add_dps_to_request(17)
    print( d.status() )
    d.updatedps([17,18,19,20])
    print( d.status() )
    d.close()

print("\nDone")
```

```
{'dps': {'1': False, '9': 0, '18': 0, '19': 0, '20': 1209, '21': 1, '22': 1123, '23': 6986, '24': 7260, '25': 4960, '26': 0}}
...
{'dps': {'1': False, '9': 0, '18': 0, '19': 0, '20': 1204, '21': 1, '22': 1136, '23': 7223, '24': 7446, '25': 4830, '26': 0}}
...
{'dps': {'1': False, '9': 0, '38': 'off', '42': '', '43': '', '44': '', '47': 'flip'}}
```

**Comment by uzlonewolf** [comment]
On further investigation it seems it was not related to add_dps or updatedps at all.  I'm still trying to figure out exactly how it works, but it seems to update automatically but only after a certain amount of time has passed.  If I just loop and .status() it every 5 minutes it will eventually report power used:
```python
import time
import tinytuya

d = tinytuya.OutletDevice( ... )

print( d.status() )
print(" > Begin Monitor Loop <")
pingtime = time.time() + 9
polltime = time.time() + 300

while(True):
    if( pingtime <= time.time() ):
        payload = d.generate_payload(tinytuya.HEART_BEAT)
        data = d.send(payload)
        pingtime = time.time() + 9
        if data:
            print( data )

    if( polltime <= time.time() ):
        polltime = time.time() + 300
        print( '================' )
        print( d.status() )

    data = d.receive()
    if data:
        print( data )
```
```
================
{'dps': {'1': False, '9': 0, '17': 24, '18': 0, '19': 0, '20': 1196, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': False, '9': 0, '17': 24, '18': 0, '19': 0, '20': 1196, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}

<repeat the '17':24 all night long, 8+ hours>

{'dps': {'1': False, '9': 0, '17': 24, '18': 0, '19': 0, '20': 1196, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': False, '9': 0, '17': 24, '18': 0, '19': 0, '20': 1196, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
{'dps': {'1': True}, 't': 1676739547} <- turned on via app.  next update has it reset to "1"
{'dps': {'20': 1200, '18': 778, '19': 935}, 't': 1676739579}
================
{'dps': {'1': True, '9': 0, '17': 1, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 1, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 1, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}


<restarted script, timer reset>

{'dps': {'1': True, '9': 0, '17': 1, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
 > Begin Monitor Loop <
================
{'dps': {'1': True, '9': 0, '17': 1, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 1, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 44, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 44, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 44, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
```
The bars are the 5-min timer triggering.  At this point I suspect a fixed 30- or 60-minute update period.  It also resets back to "1" on certain conditions that I haven't quite figured out yet, but "turn on" seems to be one of them.

I got the "TOPGREENER TGWF115PQM 4-Pack" off Amazon https://www.amazon.com/dp/B07D8ZNNPZ

> if only we had a way of categorizing all our Tuya devices and saying "loop through all devices where type=energy-plugs"

Would have been nice if they gave energy-monitoring plugs their own category code, but noooooooo, they re-used the (non-monitoring) smart plug category :(

**Comment by uzlonewolf** [comment]
It just ticked over, so it appears to be a 30-minute update period.  It also seems to reset every period as well; I have a pretty consistent 90w load connected and 45 Wh matches a 30-minute run-time.

```{'dps': {'1': True, '9': 0, '17': 1, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
 > Begin Monitor Loop <
================
{'dps': {'1': True, '9': 0, '17': 1, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 1, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 44, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 44, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 44, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 44, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 44, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 44, '18': 778, '19': 935, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
{'dps': {'18': 744, '19': 890}, 't': 1676743182}
================
{'dps': {'1': True, '9': 0, '17': 45, '18': 744, '19': 890, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
================
{'dps': {'1': True, '9': 0, '17': 45, '18': 744, '19': 890, '20': 1200, '21': 1, '22': 1020, '23': 13000, '24': 12506, '25': 2880, '26': 0, '38': 'off', '39': False, '40': 'relay', '41': False, '42': '', '43': '', '44': ''}}
```

**Comment by uzlonewolf** [comment]
It still needs to be verified, but at this point I feel the operation works like this:

When the switch is Off, DPS 17 holds the last value.
When first switched On, if it was off for at least 30 minutes then reset DPS 17 to "1"
While the switch is On, update DPS 17 every 30 minutes and reset the internal counter to "1"

**Comment by arexms** [comment]
thanks @uzlonewolf for your observations.
I'll try to do the same, but somehow I've lost a possibility to connect  to my devices via local net...
I'm receiving "Pooling 192.168... Failed:" during scanning...

But what's interesting below code works fine (via cloud):
```python
c = tinytuya.Cloud()
id = "..."
result = c.getstatus(id)
print("Status of device:\n", json.dumps(result, indent = 4))
```

And finally I'm getting 'add_ele' value so progress :)
```
Status of device:
 {
    "result": [
        {
            "code": "switch_1",
            "value": true
        },
        {
            "code": "countdown_1",
            "value": 0
        },
        {
            "code": "add_ele",
            "value": 27
        },...
```

**Comment by jasonacox** [comment]
@arexms On the local access, make sure you don't have other things connecting to the device at the same time. Many seem to be limited on the number of local connections they can support. I have devices that won't respond locally when I have the Smart Life app open.

**Comment by peterodro** [comment]
o in other words
i need to set MyTodayEnergy to 0 at midnight
and then check in a loop if dps17 is change and then add new value of dpa17 to MyTodayEnergy (except case that dps17 is change to 1)

**Comment by fhempy** [comment]
I would suggest to just check DPS 17 every 30 minutes and add the value to todays_energy only if DPS 17 is > 1. Isn't it as simple as that?

**Comment by uzlonewolf** [comment]
> Isn't it as simple as that?

No, unfortunately it's not.  If the switch is Off then DPS 17 will hold the last value even though no power is being used.  Another issue is the fact that the 30-minute timer gets reset when the switch gets turned On; if you sample too close to the edge it'll be 50/50 whether you get the old value or the new value.

**Comment by fhempy** [comment]
You are right, I forgot that the last value remains unchanged.

What about combining add_ele with cur_power? E.g. use add_ele only when cur_power was > 0 in the last 30 minutes. I just check that with my data if that assumption makes sense.

**Comment by peterodro** [comment]
i have tuya device without switch only measurement.

€ 27,69  53%OFF | Tuya Smart WiFi Electricity KWH Meter Din Rail Single Phase 110V 230V with 50A 63A CT AC Meter App Real Time Monitor Power

by observating i see it change DPS17 every 10 mins

but anyhow, if device is off (device with switch) is and keep value without change then no problem because I adding DPS17 only in case when it changes value

**Comment by peterodro** [comment]
![Image 2](https://user-images.githubusercontent.com/21343964/220715043-5e99c3c3-3959-40aa-a4aa-6880216bfff8.jpg)
screenshots..

**Comment by fhempy** [comment]
That's because within the last measuring interval the switch was not on all the time. It will remain at 71 now till you switch it on again.

**Comment by uzlonewolf** [comment]
> as see at 18:51 DPS17 from 100 to 71 while plug switched off (DPS1)

That's because the 30-minute period started at 18:21, and it was turned on from 18:21 to 18:29.

Basically the way it works is like this:  When the device is first switched On it starts a 30-minute timer.  At the end of those 30 minutes it copies the internal "power used" counter into DPS 17 and resets that internal counter to "1".  If the switch is still On it starts the timer again, otherwise it stops and waits until the switch is turned On.

**Comment by TheColector** [comment]
Hi,
Since this is not closed I will chip in my bit here.
I have recently started using Tuya Smart plug device (TS011F) over Zigbee. It's paired to [Tuya WiFi/Zigbee gateway hub](https://solution.tuya.com/projects/CMa4s3xhnq7vab) and thanks to this amazing tinytuya lib, I was able to read DPS data and control the switch, which was great because it turned out I did not need to go around and figure out the Zigbee bridging part via some other methods (like Zigbee2MQTT).

Regarding the measurement part, the plug only reports the voltage by DPS20 - any idea why it's scaled by a factor of 10?

I used the following piece of code for initial tests:

```
gw = tinytuya.Device( my_dev_id, address="my_ip", local_key='my_key', persist=True, version=3.4 )
print( 'GW IP found:', gw.address )

zigbee1 = tinytuya.OutletDevice( outlet_id, cid=outlet_cid, parent=gw )
zigbee1.set_socketPersistent( True )
zigbee1.add_dps_to_request(17)
zigbee1.updatedps([18,19,20])
print( zigbee1.status() )
```

Output I got:
- GW IP found: 192.168.0.101
- {'dps': {'1': True, '9': 0, '17': 0, '18': 0, '19': 0, '20': 2330, '27': 'off', '28': 'relay', '29': False}, 'cid': 'a4c13800218affff', 'device': OutletDevice( 'bf4055c9f8f72b3e4d0zvt', address=None, local_key='', dev_type='default', connection_timeout=5, version=3.4, persist=True, cid='a4c13800218affff', parent='bf3947a3c726d20871pkpm', children={} )}

The voltage is reported correctly, but there is not current or energy being reported (DPS[18.19].
According to [This Link](https://developer.tuya.com/en/docs/connect-subdevices-to-gateways/tuya-zigbee-measuring-smart-plug-access-standard?id=K9ik6zvofpzqk#title-15-DP19%20Power), these DPSs are meant to report that.
It's working fine via Tuya app.

@jasonacox or someone else, any ideas here?
Thanks

**Comment by make-all** [comment]
You are either not drawing current through the device or it's current measurement is broken. I guess voltage is scaled by 10 to give it more resolution while still using integers (since the microcontrollers used in these sorts of devices do not have an FPU in them to work with floating point).

**Comment by TheColector** [comment]
> You are either not drawing current through the device or it's current measurement is broken. I guess voltage is scaled by 10 to give it more resolution while still using integers (since the microcontrollers used in these sorts of devices do not have an FPU in them to work with floating point).

I double checked just a minute ago, and guess what, now it works 😆
I believe that it might have been glitched or something, because I was next to the lamp that was plugged in while I was observing the logs from python.
It was 4.5W LED garage lamp (~230V, 0.02A). After that test, I plugging in laptop charger and electric kettle after that. Only then I started to see current and power measurements.
Also, at the end, I plugged in the LED lamp from initial tests, it started showing measurements as well.

It's a mystery for me

**Comment by jasonacox-sam** [comment]
Hi @arexms and everyone who contributed here! 👋

This has been a great thread — lots of useful exploration of how energy monitoring works on Tuya plugs (DPS 17 behavior, 30-minute update cycles, cloud vs local data). To summarize: the energy consumption data shown in the app (daily/monthly kWh) is computed by the Tuya Cloud, not stored on the device itself. Some devices do expose DPS 17 with periodic incremental readings, but it varies by firmware and device type.

The question has been thoroughly answered so I'm going to close this out. If anyone has new findings or needs to revisit this, feel free to re-open or open a new issue. Thanks!

— Sam ⚙️
